src/main/java/net/neoforged/neoforge/network/registration/IPayloadRegistrar.java

```diff
@@ -0,0 +1,194 @@
+import net.neoforged.neoforge.network.handling.ITaskCompletedHandler;
+import net.neoforged.neoforge.network.handling.PlayPayloadContext;
+
+/**
+ * Defines a registrar for custom payloads that can be sent over the network.
+ * <p>
+ * A custom payload is a class which extends {@link CustomPacketPayload}, it is recommended to use a record for this.
+ * </p>
+ * <p>
+ * The payload is written to the networks outgoing buffer using the {@link CustomPacketPayload#write(FriendlyByteBuf)} method.
+ * However, to read the payload from the incoming buffer, your registered {@link FriendlyByteBuf.Reader} is used.
+ * <br>
+ * When you implement your {@link CustomPacketPayload#write(FriendlyByteBuf)} method you do not need to write the id of the payload,
+ * neither do you need to read it in your {@link FriendlyByteBuf.Reader} implementation. However, you do need to make sure that the
+ * id you pass into {@link #play(ResourceLocation, FriendlyByteBuf.Reader, IPlayPayloadHandler)} and
+ * {@link #configuration(ResourceLocation, FriendlyByteBuf.Reader, IConfigurationPayloadHandler)} is the same as the id you
+ * return from your {@link CustomPacketPayload#id()}. We suggest using a <code>public static final ResourceLocation</code> field
+ * to store it and then reference it in both places.
+ * <br>
+ * Ids can be reused between play and configuration payloads, but it is needed to use different ids for different payloads.
+ * <br>
+ * Under certain situations you are not able to register a payload:
+ * <ul>
+ * <li>If the id you are trying to register is already in use, meaning you used the same id twice for different packets of the same kind.</li>
+ * <li>If you are trying to register a payload to a namespace that is not your own.</li>
+ * </ul>
+ * In principal this means that the registration will fail if any of these cases occur.
+ * The exception thrown in these cases is a {@link RegistrationFailedException}.
+ * Although you are free to capture this exception and handle it, it is not recommended to do so, since it is a sign of a programming error.
+ * <br>
+ * </p>
+ * <p>
+ * There are two kinds of payloads:
+ * <ul>
+ * <li>Play payloads: These are payloads that are sent from the client to the server, or from the server to the client, during normal gameplay.</li>
+ * <li>Configuration payloads: These are payloads that are sent from the server to the client, or from the client to the server, during the login process, before the player is spawned.</li>
+ * </ul>
+ * You can register a custom payload for either of these types of payloads using the {@link #play(ResourceLocation, FriendlyByteBuf.Reader, IPlayPayloadHandler)}
+ * and {@link #configuration(ResourceLocation, FriendlyByteBuf.Reader, IConfigurationPayloadHandler)} methods respectively.
+ * <br>
+ * The difference between the play and configuration phases, if you like to call them that, is that the configuration phase generally requires
```

Comment:
Javadocs should not refer to the reader or use expressions such as `if you'd like to call them that`. Use proper names whose meaning is documented.

Reply:
Documentation on the styling of JDoc guidelines for Forge?
This JDoc is a verbatim description of the core registrar and is meant as documentation for novices and experts alike. I see no reason to change the wording.